File: native/android/ProtocolModels.kt
package remux.protocol

data class ProtocolCapabilities(
    val envelope: Boolean,
    val inspectV2: Boolean,
    val deviceTrust: Boolean,
)

data class AuthPayload(
    val token: String,
    val password: String?,
    val cols: Int?,
    val rows: Int?,
    val capabilities: ProtocolCapabilities?,
)

data class AuthOkPayload(
    val capabilities: ProtocolCapabilities,
)

data class AuthErrorPayload(
    val reason: String,
)

data class InspectContentPayload(
    val content: String,
)

data class RemuxEnvelope<T>(
    val domain: String,
    val type: String,
    val version: Int,
    val requestId: String?,
    val emittedAt: String,
    val source: String,
    val payload: T,
)

data class WorkspacePane(
    val id: String,
    val focused: Boolean,
    val title: String,
    val command: String?,
    val cwd: String?,
    val rows: Int,
    val cols: Int,
    val x: Int,
    val y: Int,
)

data class WorkspaceTab(
    val index: Int,
    val name: String,
    val active: Boolean,
    val isFullscreen: Boolean,
    val hasBell: Boolean,
    val panes: List<WorkspacePane>,
)

data class WorkspaceState(
    val session: String,
    val tabs: List<WorkspaceTab>,
    val activeTabIndex: Int,
)

data class InspectHighlight(
    val start: Int,
    val end: Int,
)

data class InspectDescriptor(
    val scope: String,
    val source: String,
    val precision: String,
    val staleness: String,
    val capturedAt: String,
    val paneId: String?,
    val tabIndex: Int?,
    val totalItems: Int?,
)

data class InspectItem(
    val type: String,
    val content: String,
    val lineNumber: Int?,
    val timestamp: String,
    val paneId: String?,
    val highlights: List<InspectHighlight>?,
)

data class InspectSnapshot(
    val descriptor: InspectDescriptor,
    val items: List<InspectItem>,
    val cursor: String?,
    val truncated: Boolean,
)

data class InspectRequest(
    val scope: String,
    val paneId: String?,
    val tabIndex: Int?,
    val cursor: String?,
    val query: String?,
    val limit: Int?,
)

data class BandwidthStats(
    val rawBytesPerSec: Double,
    val compressedBytesPerSec: Double,
    val savedPercent: Double,
    val fullSnapshotsSent: Int,
    val diffUpdatesSent: Int,
    val avgChangedRowsPerDiff: Double,
    val totalRawBytes: Int,
    val totalCompressedBytes: Int,
    val totalSavedBytes: Int,
    val rttMs: Int?,
    val protocolName: String,
)

data class BandwidthStatsPayload(
    val stats: BandwidthStats,
)

data class LegacyAuthOk(
    val type: String,
    val capabilities: ProtocolCapabilities,
)

data class LegacyAuth(
    val type: String,
    val token: String,
    val password: String?,
    val cols: Int?,
    val rows: Int?,
    val capabilities: ProtocolCapabilities?,
)

data class LegacyAuthError(
    val type: String,
    val reason: String,
)

data class LegacyErrorMessage(
    val type: String,
    val code: Int?,
    val message: String,
)

data class LegacyPong(
    val type: String,
    val timestamp: Double,
)

data class LegacyWorkspaceState(
    val type: String,
    val session: String,
    val tabs: List<WorkspaceTab>,
    val activeTabIndex: Int,
)

data class LegacyInspectRequest(
    val type: String,
    val scope: String,
    val paneId: String?,
    val tabIndex: Int?,
    val cursor: String?,
    val query: String?,
    val limit: Int?,
)

data class LegacyInspectSnapshot(
    val type: String,
    val descriptor: InspectDescriptor,
    val items: List<InspectItem>,
    val cursor: String?,
    val truncated: Boolean,
)

data class LegacyBandwidthStats(
    val type: String,
    val stats: BandwidthStats,
)

data class LegacyInspectContent(
    val type: String,
    val content: String,
)

File: labs/protocol-goldens/native/ios/ProtocolModels.swift
public struct ProtocolCapabilities: Codable, Equatable {
  public let envelope: Bool
  public let inspectV2: Bool
  public let deviceTrust: Bool
}

public struct AuthPayload: Codable, Equatable {
  public let token: String
  public let password: String?
  public let cols: Int?
  public let rows: Int?
  public let capabilities: ProtocolCapabilities?
}

public struct AuthOkPayload: Codable, Equatable {
  public let capabilities: ProtocolCapabilities
}

public struct AuthErrorPayload: Codable, Equatable {
  public let reason: String
}

public struct InspectContentPayload: Codable, Equatable {
  public let content: String
}

public struct RemuxEnvelope<Payload: Codable & Equatable>: Codable, Equatable {
  public let domain: String
  public let type: String
  public let version: Int
  public let requestId: String?
  public let emittedAt: String
  public let source: String
  public let payload: Payload
}

public struct WorkspacePane: Codable, Equatable {
  public let id: String
  public let focused: Bool
  public let title: String
  public let command: String?
  public let cwd: String?
  public let rows: Int
  public let cols: Int
  public let x: Int
  public let y: Int
}

public struct WorkspaceTab: Codable, Equatable {
  public let index: Int
  public let name: String
  public let active: Bool
  public let isFullscreen: Bool
  public let hasBell: Bool
  public let panes: [WorkspacePane]
}

public struct WorkspaceState: Codable, Equatable {
  public let session: String
  public let tabs: [WorkspaceTab]
  public let activeTabIndex: Int
}

public struct InspectHighlight: Codable, Equatable {
  public let start: Int
  public let end: Int
}

public struct InspectDescriptor: Codable, Equatable {
  public let scope: String
  public let source: String
  public let precision: String
  public let staleness: String
  public let capturedAt: String
  public let paneId: String?
  public let tabIndex: Int?
  public let totalItems: Int?
}

public struct InspectItem: Codable, Equatable {
  public let type: String
  public let content: String
  public let lineNumber: Int?
  public let timestamp: String
  public let paneId: String?
  public let highlights: [InspectHighlight]?
}

public struct InspectSnapshot: Codable, Equatable {
  public let descriptor: InspectDescriptor
  public let items: [InspectItem]
  public let cursor: String?
  public let truncated: Bool
}

public struct InspectRequest: Codable, Equatable {
  public let scope: String
  public let paneId: String?
  public let tabIndex: Int?
  public let cursor: String?
  public let query: String?
  public let limit: Int?
}

public struct BandwidthStats: Codable, Equatable {
  public let rawBytesPerSec: Double
  public let compressedBytesPerSec: Double
  public let savedPercent: Double
  public let fullSnapshotsSent: Int
  public let diffUpdatesSent: Int
  public let avgChangedRowsPerDiff: Double
  public let totalRawBytes: Int
  public let totalCompressedBytes: Int
  public let totalSavedBytes: Int
  public let rttMs: Int?
  public let protocolName: String

  enum CodingKeys: String, CodingKey {
    case rawBytesPerSec
    case compressedBytesPerSec
    case savedPercent
    case fullSnapshotsSent
    case diffUpdatesSent
    case avgChangedRowsPerDiff
    case totalRawBytes
    case totalCompressedBytes
    case totalSavedBytes
    case rttMs
    case protocolName = "protocol"
  }
}

public struct BandwidthStatsPayload: Codable, Equatable {
  public let stats: BandwidthStats
}

public struct LegacyAuthOk: Codable, Equatable {
  public let type: String
  public let capabilities: ProtocolCapabilities
}

public struct LegacyAuth: Codable, Equatable {
  public let type: String
  public let token: String
  public let password: String?
  public let cols: Int?
  public let rows: Int?
  public let capabilities: ProtocolCapabilities?
}

public struct LegacyAuthError: Codable, Equatable {
  public let type: String
  public let reason: String
}

public struct LegacyErrorMessage: Codable, Equatable {
  public let type: String
  public let code: Int?
  public let message: String
}

public struct LegacyPong: Codable, Equatable {
  public let type: String
  public let timestamp: Double
}

public struct LegacyWorkspaceState: Codable, Equatable {
  public let type: String
  public let session: String
  public let tabs: [WorkspaceTab]
  public let activeTabIndex: Int
}

public struct LegacyInspectRequest: Codable, Equatable {
  public let type: String
  public let scope: String
  public let paneId: String?
  public let tabIndex: Int?
  public let cursor: String?
  public let query: String?
  public let limit: Int?
}

public struct LegacyInspectSnapshot: Codable, Equatable {
  public let type: String
  public let descriptor: InspectDescriptor
  public let items: [InspectItem]
  public let cursor: String?
  public let truncated: Bool
}

public struct LegacyBandwidthStats: Codable, Equatable {
  public let type: String
  public let stats: BandwidthStats
}

public struct LegacyInspectContent: Codable, Equatable {
  public let type: String
  public let content: String
}
